src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -0,0 +1,161 @@
+  /// \param reply gRPC Reply.
+  /// \param send_reply_callback Callback to invoke when sending reply.
+  void HandleAddTaskEventData(rpc::AddTaskEventDataRequest request,
+                              rpc::AddTaskEventDataReply *reply,
+                              rpc::SendReplyCallback send_reply_callback) override;
+
+  ///  Handles a GetTaskEvents request.
+  ///
+  /// \param request gRPC Request.
+  /// \param reply gRPC Reply.
+  /// \param send_reply_callback Callback to invoke when sending reply.
+  void HandleGetTaskEvents(rpc::GetTaskEventsRequest request,
+                           rpc::GetTaskEventsReply *reply,
+                           rpc::SendReplyCallback send_reply_callback) override {
+    RAY_LOG(FATAL) << "Unimplemented!";
+  };
+
+  /// Stops the event loop and the thread of the task event handler.
+  ///
+  /// This function returns when the io thread is joined.
+  void Stop();
+
+  /// Returns the io_service.
+  ///
+  /// \return Reference to its io_service.
+  instrumented_io_context &GetIoContext() { return io_service_; }
+
+  /// A storage component that stores the task events.
+  ///
+  /// This is an in-memory storage component that supports adding and getting of task
+  /// events.
+  ///
+  /// It merges events from a single task attempt (same task id and attempt number) into
+  /// a single rpc::TaskEvents entry, as reported by multiple rpc calls from workers.
+  ///
+  /// When more than `RAY_task_events_max_num_task_in_gcs` task events are stored in the
+  /// the storage, older task events will be replaced by new task events, where older
+  /// task events are approximately task events that arrived in earlier rpc.
+  class GcsTaskManagerStorage {
+   public:
+    /// Constructor
+    ///
+    /// \param max_num_task_events Max number of task events stored before replacing older
+    /// ones.
+    GcsTaskManagerStorage(size_t max_num_task_events)
+        : max_num_task_events_(max_num_task_events) {}
+
+    /// Add a new task event or replace an existing task event in the storage.
+    ///
+    /// If there are already `RAY_task_events_max_num_task_in_gcs` in the storage, the
+    /// oldest task event will be replaced. Otherwise the `task_event` will be added.
+    ///
+    /// \param task_event Task event to be added to the storage.
+    /// \return absl::nullptr if the `task_event` is added without replacement, else the
+    /// replaced task event.
+    absl::optional<rpc::TaskEvents> AddOrReplaceTaskEvent(rpc::TaskEvents task_event);
+
+    /// Get task events.
+    ///
+    /// \param job_id Getting task events from this `job_id` only.
+    /// \return A vector of task events.
+    std::vector<rpc::TaskEvents> GetTaskEvents(
+        absl::optional<JobID> job_id = absl::nullopt) = delete;
+
+    /// Max number of task events allowed in the storage.
+    const size_t max_num_task_events_ = 0;
+
+    /// Current task events stored.
+    std::vector<rpc::TaskEvents> task_events_;
```

Comment:
Fine for now, but if the logic gets more complicated we should consider splitting this out into a LRU data struture.